src/lib/core/OTAImageHeader.cpp

```diff
@@ -39,23 +39,88 @@ enum class Tag : uint8_t
     kImageDigest           = 9,
 };
 
+/// Length of the fixed portion of the Matter OTA image header: FileIdentifier (4B), TotalSize (8B) and HeaderSize (4B)
+constexpr uint32_t kFixedHeaderSize = 16;
+
+/// Maximum size of the software version string
+constexpr size_t kMaxSoftwareVersionStringSize = 64;
+
+/// Maximum size of the release notes URL
+constexpr size_t kMaxReleaseNotesURLSize = 256;
+
 } // namespace
 
-CHIP_ERROR DecodeOTAImageHeader(ByteSpan buffer, OTAImageHeader & header)
+void OTAImageHeaderParser::Init()
+{
+    mState         = State::kInitialized;
+    mBufferOffset  = 0;
+    mHeaderTlvSize = 0;
+    mBuffer.Alloc(kFixedHeaderSize);
+}
+
+void OTAImageHeaderParser::Clear()
+{
+    mState         = State::kNotInitialized;
+    mBufferOffset  = 0;
+    mHeaderTlvSize = 0;
+    mBuffer.Free();
+}
+
+CHIP_ERROR OTAImageHeaderParser::AccumulateAndDecode(ByteSpan & buffer, OTAImageHeader & header)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    if (mState == State::kInitialized)
+    {
+        Append(buffer, kFixedHeaderSize - mBufferOffset);
+        error = DecodeFixed();
+    }
+
+    if (mState == State::kTlv)
+    {
+        Append(buffer, mHeaderTlvSize - mBufferOffset);
+        error = DecodeTlv(header);
+    }
+
+    if (error != CHIP_NO_ERROR && error != CHIP_ERROR_BUFFER_TOO_SMALL)
+    {
+        Clear();
+    }
+
+    return error;
+}
+
+void OTAImageHeaderParser::Append(ByteSpan & buffer, uint32_t numBytes)
 {
-    Encoding::LittleEndian::Reader reader(buffer);
+    numBytes = chip::min(numBytes, static_cast<uint32_t>(buffer.size()));
+    memcpy(&mBuffer[mBufferOffset], buffer.data(), numBytes);
+    mBufferOffset += numBytes;
+    buffer = buffer.SubSpan(numBytes);
+}
 
-    // Parse the fixed part of the header
+CHIP_ERROR OTAImageHeaderParser::DecodeFixed()
+{
+    ReturnErrorCodeIf(mBufferOffset < kFixedHeaderSize, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    Encoding::LittleEndian::Reader reader(mBuffer.Get(), mBufferOffset);
     uint32_t fileIdentifier;
-    ReturnErrorOnFailure(reader.Read32(&fileIdentifier).StatusCode());
-    VerifyOrReturnError(fileIdentifier == kOTAImageFileIdentifier, CHIP_ERROR_INVALID_FILE_IDENTIFIER);
-    ReturnErrorOnFailure(reader.Read64(&header.mTotalSize).Read32(&header.mHeaderSize).StatusCode());
+    uint64_t totalSize;
+    ReturnErrorOnFailure(reader.Read32(&fileIdentifier).Read64(&totalSize).Read32(&mHeaderTlvSize).StatusCode());
+    ReturnErrorCodeIf(fileIdentifier != kOTAImageFileIdentifier, CHIP_ERROR_INVALID_FILE_IDENTIFIER);
+    ReturnErrorCodeIf(!mBuffer.Alloc(mHeaderTlvSize), CHIP_ERROR_NO_MEMORY);
```

Comment:
This code is extremely dangerous, as it forces a heap allocation based on trusting a remotely-provided size, without any bounds checks. There is an absolute maximum size that the TLV header could be: 499 bytes. To account for future changes and updates, I think there needs to be a validation that the OTA header size <= 1k.

Test code for max size, based on TLVWriter python class (src/controller/python/chip/tlv/__init__.py):

```
    writer = TLVWriter()
    val = {
        0: 0xFFFF, #VendorID              [0] : UNSIGNED INTEGER [ range 16bits ],
        1: 0xFFFF, # ProductID             [1] : UNSIGNED INTEGER [ range 16bits ],
        2: 0xFFFFFFFF, # SoftwareVersion       [2] : UNSIGNED INTEGER [ range 32bits ],
        3: "a" * 64, # SoftwareVersionString [3] : STRING [ length 1..64 ],
        4: 0xFFFFFFFF_FFFFFFFF, # PayloadSize           [4] : UNSIGNED INTEGER [ range 64bits ],
        5: 0xFFFF_FFFF, # MinApplicableSoftwareVersion    [5, optional] : UNSIGNED INTEGER [ range 32bits ],
        6: 0xFFFF_FFFF, # MaxApplicableSoftwareVersion    [6, optional] : UNSIGNED INTEGER [ range 32bits ],
        7: "a" * 256, # ReleaseNotesURL       [7, optional] : STRING [ length 1..256 ],
        8: 0xFF, # ImageDigestType       [8] : UNSIGNED INTEGER [ range 8bits ],
        9: "\xaa" * 64 # ImageDigest           [9] : OCTET STRING [ length 0..64 ]
    }

    writer.put(None,val)
    out = writer.encoding
    print("TLV len: %d, out: %s" % (len(out), out))

------------

TLV len: 499, out: bytearray(b"\x15%\x00\xff\xff%\x01\xff\xff&\x02\xff\xff\xff\xff,\x03@aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\'\x04\xff\xff\xff\xff\xff\xff\xff\xff&\x05\xff\xff\xff\xff&\x06\xff\xff\xff\xff-\x07\x00\x01aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa$\x08\xff,\t\x80\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\xc2\xaa\x18")
```

Reply:
Right, good point.

Reply:
Added a safety check.